evalml/pipelines/components/transformers/preprocessing/stl_decomposer.py

```diff
@@ -0,0 +1,384 @@
+            right_index = (
+                y_t.index[-1] + 1
+                if isinstance(y_t.index, (Int64Index, pd.RangeIndex))
+                else y_t.index[-1] + 1 * y_t.index.freq
+            )
+            trend = (
+                self.trend.reset_index(drop=True)[left_index:right_index]
+                if isinstance(y_t.index, (Int64Index, pd.RangeIndex))
+                else self.trend[left_index:right_index]
+            )
+            seasonal = (
+                self.seasonal.reset_index(drop=True)[left_index:right_index]
+                if isinstance(y_t.index, (Int64Index, pd.RangeIndex))
+                else self.seasonal[left_index:right_index]
+            )
+            y_in_sample = y_t + trend + seasonal
+            y_in_sample = y_in_sample.dropna()
+
+        # For out of sample data....
+        if y_t.index[-1] > index[-1]:
+            try:
+                # ...that is partially out of sample and partially in sample.
+                truncated_y_t = y_t[y_t.index.get_loc(index[-1]) + 1 :]
+            except KeyError:
+                # ...that is entirely out of sample.
+                truncated_y_t = y_t
+            (
+                projected_trend,
+                projected_seasonality,
+            ) = self._project_trend_and_seasonality(truncated_y_t)
+
+            y_out_of_sample = infer_feature_types(
+                pd.Series(
+                    truncated_y_t + projected_trend + projected_seasonality,
+                    index=truncated_y_t.index,
+                ),
+            )
+        y = y_in_sample.append(y_out_of_sample)
+        y.index = original_index
+        return y
+
+    def get_trend_dataframe(self, X, y):
+        """Return a list of dataframes with 4 columns: signal, trend, seasonality, residual.
+
+        Args:
+            X (pd.DataFrame): Input data with time series data in index.
+            y (pd.Series or pd.DataFrame): Target variable data provided as a Series for univariate problems or
+                a DataFrame for multivariate problems.
+
+        Returns:
+            list of pd.DataFrame: Each DataFrame contains the columns "signal", "trend", "seasonality" and "residual,"
+                with the latter 3 column values being the decomposed elements of the target data.  The "signal" column
+                is simply the input target signal but reindexed with a datetime index to match the input features.
+
+        Raises:
+            TypeError: If X does not have time-series data in the index.
+            ValueError: If time series index of X does not have an inferred frequency.
+            ValueError: If the forecaster associated with the detrender has not been fit yet.
+            TypeError: If y is not provided as a pandas Series or DataFrame.
+
+        """
+        X = infer_feature_types(X)
+        if not isinstance(X.index, pd.DatetimeIndex):
+            raise TypeError("Provided X should have datetimes in the index.")
```

Comment:
Do we need to continue enforcing this? We can set it within this function relatively easily.